src/hotspot/share/opto/addnode.cpp

```diff
@@ -1086,166 +1086,98 @@ static bool can_overflow(const TypeInt* t, jint c) {
-      y->in(2)->is_Con()) {
-    const Type* t = y->in(2)->bottom_type();
-    if (t == Type::TOP) return nullptr;  // No progress
-    y_off = t->is_int()->get_con();
-    y = y->in(1);
-  }
-  if (x->_idx > y->_idx && r->Opcode() != Op_MaxI) {
-    swap_edges(1, 2);
-    return this;
+// Return:
+// <x, C>,       if n is of the form x + C, where 'C' is a non-TOP constant;
+// <nullptr, _>, if n is of the form x + C, where 'C' is a TOP constant;
+// <n, con>      otherwise.
+static Node* constant_add_input(Node* n, jint* con) {
+  if (n->Opcode() == Op_AddI && n->in(2)->is_Con()) {
+    const Type* t = n->in(2)->bottom_type();
+    if (t == Type::TOP) {
+      return nullptr;
+    }
+    *con = t->is_int()->get_con();
+    n = n->in(1);
   }
+  return n;
+}
 
+// Transform opcode(x + x_off, y + y_off) into x + opcode(x_off, y_off), where
+// opcode is either MinI or MaxI, if x == y and the additions can't overflow.
+static Node* extract_addition(PhaseGVN* phase, Node* x, jint x_off, Node* y, jint y_off, int opcode) {
+  assert(opcode == Op_MaxI || opcode == Op_MinI, "Unexpected opcode");
   const TypeInt* tx = phase->type(x)->isa_int();
+  if (x == y && tx != nullptr &&
+      !can_overflow(tx, x_off) &&
+      !can_overflow(tx, y_off)) {
+    jint c = opcode == Op_MinI ? MIN2(x_off, y_off) : MAX2(x_off, y_off);
+    return new AddINode(x, phase->intcon(c));
+  }
+  return nullptr;
+}
 
-  if (r->Opcode() == Op_MaxI) {
-    assert(r != r->in(2), "dead loop in MaxINode::Ideal");
-    y = r->in(1);
-    // Check final part of MAX tree
-    if (y->Opcode() == Op_AddI &&// Check for "y+c1" and collect constant
-        y->in(2)->is_Con()) {
-      const Type* t = y->in(2)->bottom_type();
-      if (t == Type::TOP) return nullptr;  // No progress
-      y_off = t->is_int()->get_con();
-      y = y->in(1);
-    }
-
-    if (x->_idx > y->_idx)
-      return new MaxINode(r->in(1), phase->transform(new MaxINode(l, r->in(2))));
-
-    // Transform MAX2(x + c0, MAX2(x + c1, z)) into MAX2(x + MAX2(c0, c1), z)
-    // if x == y and the additions can't overflow.
-    if (x == y && tx != nullptr &&
-        !can_overflow(tx, x_off) &&
-        !can_overflow(tx, y_off)) {
-      return new MaxINode(phase->transform(new AddINode(x, phase->intcon(MAX2(x_off, y_off)))), r->in(2));
+Node* MaxNode::IdealI(PhaseGVN* phase, bool can_reshape, int opcode) {
+  assert(opcode == Op_MinI || opcode == Op_MaxI, "Unexpected opcode");
+  Node* l = in(1);
+  Node* r = in(2);
+  jint x_off = 0;
+  Node* x = constant_add_input(l, &x_off);
+  if (x == nullptr) return nullptr;
+  jint y_off = 0;
+  Node* y = constant_add_input(r, &y_off);
+  if (y == nullptr) return nullptr;
+  // Try to transform opcode(x + x_off, opcode(y + y_off, z)) (in any of its
+  // four possible permutations given by opcode's commutativity) into
+  // opcode(x + opcode(x_off, y_off), z), where opcode is either MinI or MaxI,
+  // if x == y and the additions can't overflow.
```

Comment:
Ok, effectively we have 5, not just 4 cases here:
```
opcode(x + x_off, opcode(y + y_off, z))
opcode(x + x_off, opcode(z, y + y_off))
opcode(opcode(y + y_off, z), x + x_off)
opcode(opcode(z, y + y_off), x + x_off)
opcode(x + x_off, y + y_off)
```

I find the nested for-loop quite confusing. Maybe packing the inner stuff into a separate function could work?

Maybe doing it explicit could also do the trick?
```
// Check for opcode(x + x_con,  y + y_con), no z
if (in(1)->Opcode() == Op_AddI && in(2)->Opcode() == Op_AddI) {
  Node* ret = try_fold(opcode, in(1), in(2), nullptr);
  if (ret != nullptr) { return ret; }
}

// Check for these 4 cases, equivalent to opcode3(addx, addy, z)
// opcode(x + x_con, opcode(y + y_con, z))
// opcode(x + x_con, opcode(z, y + y_con))
// opcode(opcode(y + y_con, z), x + x_con)
// opcode(opcode(z, y + y_con), x + x_con)
for (uint i = 1; i < 2; i++) {
  Node* addx = in(i);
  Node* other = in(i == 1 ? 2 : 1); // or just "2-i"
  if (addx->Opcode() != Op_AddI || other->Opcode() != opcode) { continue; }
  for (uint j = 1; i < 2; j++) {
    Node* addy = other->in(j);
    Node* z = other->in(j == 1 ? 2 : 1);
    if (addy->Opcode() != Op_AddI) { continue; }
    // We have opcode3(addx, addy, z)
    Node* ret = try_fold(opcode, addx, addy, z);
    if (ret != nullptr) { return ret; }
  }
}
```
Where we have
```
Node* try_fold(int opcode, Node* addx, Node* addy, Node* z = nullptr) {
  jint addx_con = 0;
  jint addy_con = 0;
  Node* addx_var = as_add_constant(addx, &addx_con);
  Node* addy_var = as_add_constant(addy, &addy_con);
  if (addx_var == nullptr || addy_var == nullptr) {
    // found a top
    return nullptr;
  }
  // could even check addx_var != addy_var, then we don't have to do that inside...
  Node* folded = extract_addition(phase, addx_var, addx_con, addy_var, addy_con, opcode);
  if (z != nullptr) {
    folded = opcode(folded, z);
  }
  return folded;
}
```

Maybe this does a few more calls to `as_add_constant` than strictly necessary, but it is a bit easier to understand, right?

Reply:
Thanks for the feedback, I agree the first refactor was not as clear as it should be, possibly because I derived it quite mechanically from the existing code. I have reworked `MaxNode::IdealI()` using more explicit naming, a functional and simplified version of `as_add_with_constant()`, and a clearer separation of the different optimizations that are applied. Hope it is more readable now.